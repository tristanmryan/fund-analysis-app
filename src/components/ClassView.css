.benchmark-banner {
  background: #f1f5f9; /* slate-100 */
  border-radius: 0.5rem;
  padding: 0.5rem 0.75rem;
  margin-bottom: 0.25rem;
  font-weight: 500;
}

