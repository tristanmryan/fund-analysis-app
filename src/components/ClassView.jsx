export { default } from '../routes/ClassView'
