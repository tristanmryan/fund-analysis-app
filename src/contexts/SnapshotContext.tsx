import React, { createContext, useContext, ReactNode, useCallback } from 'react'
import { liveQuery } from 'dexie'
import { useObservable } from 'dexie-react-hooks'
import db, {
  getActiveSnapshot,
  setActiveSnapshot,
  SnapshotRow
} from '../services/snapshotStore'

interface SnapshotCtxShape {
  active: SnapshotRow | undefined
  setActive: (id: string) => Promise<void>
  list: SnapshotRow[]
}

const SnapshotContext = createContext<SnapshotCtxShape | undefined>(undefined)

export function SnapshotProvider ({ children }: { children: ReactNode }) {
  const active = useObservable<SnapshotRow | undefined>(() =>
    liveQuery(getActiveSnapshot)
  )
  const list = useObservable<SnapshotRow[], SnapshotRow[]>(
    () =>
      liveQuery(() =>
        db.snapshots
          .filter(row => row.deleted !== true)
          .toArray()
      ),
    [],
    []
  )

  const setActive = useCallback(async (id: string) => {
    await setActiveSnapshot(id)
  }, [])

  return (
    <SnapshotContext.Provider value={{ active, setActive, list }}>
      {children}
    </SnapshotContext.Provider>
  )
}

export function useSnapshot () {
  const ctx = useContext(SnapshotContext)
  if (!ctx) throw new Error('useSnapshot must be used inside SnapshotProvider')
  return ctx
}
