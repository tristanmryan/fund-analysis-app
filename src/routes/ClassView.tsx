import React, { useMemo } from 'react'
import FundTable from '../components/FundTable.jsx'
import { useSnapshot } from '../contexts/SnapshotContext'
import { NormalisedRow } from '../utils/parseFundFile'

interface Props {
  defaultAssetClass?: string
}

export default function ClassView ({ defaultAssetClass }: Props) {
  const { active } = useSnapshot()
  const rows: NormalisedRow[] = active?.rows ?? []

  const funds = useMemo(() => {
    return (rows as any[]).filter(f =>
      defaultAssetClass ? (f as any).assetClass === defaultAssetClass : true
    )
  }, [rows, defaultAssetClass])

  if (!active) {
    return <div>No snapshot selected</div>
  }

  const benchmark = funds.find(f => (f as any).isBenchmark)
  const peers = funds
    .filter(f => !(f as any).isBenchmark)
    .sort((a, b) => ((b as any).scores?.final || 0) - ((a as any).scores?.final || 0))

  return (
    <div className="class-view">
      <FundTable rows={peers as any} benchmark={benchmark as any} />
    </div>
  )
}
