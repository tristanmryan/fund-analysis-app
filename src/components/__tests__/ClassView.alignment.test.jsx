import 'fake-indexeddb/auto';
import { render } from '@testing-library/react';
import ClassView from '../ClassView.jsx';
import { useSnapshot } from '../../contexts/SnapshotContext';

jest.mock('../../contexts/SnapshotContext', () => ({
  useSnapshot: jest.fn()
}));

global.structuredClone =
  global.structuredClone || ((v) => JSON.parse(JSON.stringify(v)));

const funds = [
  { Symbol: 'IWF', fundName: 'Index', isBenchmark: true, scores: { final: 60 } },
  { Symbol: 'AAA', fundName: 'Fund A', scores: { final: 70 } }
];


test('benchmark row aligns with table', () => {
  useSnapshot.mockReturnValue({ active: { rows: funds }, setActive: jest.fn(), list: [] });
  const { asFragment } = render(
    <ClassView defaultAssetClass="Unknown" />
  );
  expect(asFragment()).toMatchSnapshot();
});
