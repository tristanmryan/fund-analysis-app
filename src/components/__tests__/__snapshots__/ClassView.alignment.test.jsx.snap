// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`benchmark row aligns with table 1`] = `
<DocumentFragment>
  <div
    class="class-view"
  >
    <div
      style="overflow-x: auto;"
    >
      <table
        style="width: 100%; border-collapse: collapse;"
      >
        <colgroup>
          <col />
          <col />
          <col />
          <col />
          <col />
          <col />
          <col />
          <col />
          <col />
          <col />
          <col />
          <col />
          <col />
          <col />
        </colgroup>
        <thead>
          <tr
            style="border-bottom: 2px solid #e5e7eb;"
          >
            <th
              style="padding: 0.75rem; text-align: left; font-weight: 500; cursor: pointer;"
            >
              Symbol
            </th>
            <th
              style="padding: 0.75rem; text-align: left; font-weight: 500; cursor: pointer;"
            >
              Fund Name
            </th>
            <th
              style="padding: 0.75rem; text-align: left; font-weight: 500; cursor: pointer;"
            >
              Type
            </th>
            <th
              style="padding: 0.75rem; text-align: right; font-weight: 500; cursor: pointer;"
            >
              Score
            </th>
            <th
              style="padding: 0.75rem; text-align: right; font-weight: 500; cursor: pointer;"
            >
              Î”
            </th>
            <th
              style="padding: 0.75rem; text-align: left; font-weight: 500; cursor: pointer;"
            >
              Trend
            </th>
            <th
              style="padding: 0.75rem; text-align: right; font-weight: 500; cursor: pointer;"
            >
              YTD
            </th>
            <th
              style="padding: 0.75rem; text-align: right; font-weight: 500; cursor: pointer;"
            >
              1Y
            </th>
            <th
              style="padding: 0.75rem; text-align: right; font-weight: 500; cursor: pointer;"
            >
              3Y
            </th>
            <th
              style="padding: 0.75rem; text-align: right; font-weight: 500; cursor: pointer;"
            >
              5Y
            </th>
            <th
              style="padding: 0.75rem; text-align: right; font-weight: 500; cursor: pointer;"
            >
              Sharpe
            </th>
            <th
              style="padding: 0.75rem; text-align: right; font-weight: 500; cursor: pointer;"
            >
              Std Dev (5Y)
            </th>
            <th
              style="padding: 0.75rem; text-align: right; font-weight: 500; cursor: pointer;"
            >
              Expense
            </th>
            <th
              style="padding: 0.75rem; text-align: left; font-weight: 500; cursor: pointer;"
            >
              Tags
            </th>
          </tr>
        </thead>
        <tbody />
      </table>
    </div>
  </div>
</DocumentFragment>
`;
